lib/logstash/outputs/elasticsearch.rb

```diff
@@ -428,6 +431,14 @@ def common_event_params(event)
       params[:pipeline] = value unless value.empty?
     end
 
+    if @use_metadata
+      params[:_id] = event.get("[@metadata][_id]") || params[:_id]
+      event_index = event.get("[@metadata][_index]")
+      params[:_index] = event.sprintf(event_index) if event_index && !event_index.empty?
+      event_pipeline = event.get("[@metadata][pipeline]")
+      params[:pipeline] = event.sprintf(event_pipeline) if event_pipeline && !event_pipeline.empty?
```

Comment:
The specs don't have any coverage for sprintf-ing the value in `[@metadata][_index]`, and I personally see this as a liability since `Event#sprintf` will does not support default values -- when a en event _does_ include an `[@metadata][_event]` with a value like `logstash-%{missing}` and does not have a value for that referenced field, we end up setting the event's index to the literal `logstash-%{missing}` which can be problematic. I would rather _not_ support sprintf's and have a more constrained surface-area:

```suggestion
      event_index = event.get("[@metadata][_index]")
      params[:_index] = event_index if event_index && !event_index.empty?
      event_pipeline = event.get("[@metadata][pipeline]")
      params[:pipeline] = event_pipeline if event_pipeline && !event_pipeline.empty?
```